IPython/core/completer.py

```diff
@@ -839,22 +839,30 @@ def _default_arguments(self, obj):
             # for all others, check if they are __call__able
             elif hasattr(obj, '__call__'):
                 call_obj = obj.__call__
-
         ret += self._default_arguments_from_docstring(
                  getattr(call_obj, '__doc__', ''))
 
-        try:
-            args,_,_1,defaults = inspect.getargspec(call_obj)
-            if defaults:
-                ret+=args[-len(defaults):]
-        except TypeError:
-            pass
-
+        if PY3:
+            _keepers = (inspect.Parameter.KEYWORD_ONLY,
+                        inspect.Parameter.POSITIONAL_OR_KEYWORD)
+            try:
+                sig = inspect.signature(call_obj)
+                ret.extend(k for k, v in sig.parameters.items() if
+                           v.kind in _keepers)
+            except ValueError:
+                pass
+        else:
+            try:
+                args, _, _1, defaults = inspect.getargspec(call_obj)
```

Comment:
I wonder what would happen if on py2, this could a) unwrap (signature does unwrap until a "siganture" is found, getargspec does not) and b) check for a simple `get_additional_completions(...)` function which we could then set in out decorator. `getdoc()` already does something similar...

so:

```
addons = []
while true:
    if hasattr(call_obj, "get_additional_completions"):
        addons.extend(call_obj.get_additional_completitions())
        break
    if hasattr(call_obj, "__wrapped__"):
         call_obj = call_obj.__wrapped__
    else:
         break
[...]
ret.extend(addons)
```